fdbserver/EncryptKeyProxy.actor.cpp

```diff
@@ -654,6 +680,66 @@ bool isBlobMetadataEligibleForRefresh(const BlobMetadataDetailsRef& blobMetadata
 	return nextRefreshCycleTS > blobMetadata.expireAt || nextRefreshCycleTS > blobMetadata.refreshAt;
 }
 
+ACTOR Future<KMSHealthStatus> getHealthStatusImpl(Reference<EncryptKeyProxyData> ekpProxyData,
+                                                  KmsConnectorInterface kmsConnectorInf) {
+	state UID debugId = deterministicRandom()->randomUniqueID();
+	if (DEBUG_ENCRYPT_KEY_PROXY) {
+		TraceEvent("KMSHealthCheckStart", ekpProxyData->myId);
+	}
+
+	// Health check will try to fetch the encryption details for the system key
+	try {
+		KmsConnLookupEKsByDomainIdsReq req;
+		req.debugId = debugId;
+		req.encryptDomainIds.push_back(SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID);
+		++ekpProxyData->numHealthCheckRequests;
+		KmsConnLookupEKsByDomainIdsRep rep = wait(timeoutError(kmsConnectorInf.ekLookupByDomainIds.getReply(req),
+		                                                       FLOW_KNOBS->EKP_HEALTH_CHECK_REQUEST_TIMEOUT));
+		if (rep.cipherKeyDetails.size() < 1) {
+			TraceEvent(SevWarn, "KMSHealthCheckResponseEmpty");
+			throw encrypt_keys_fetch_failed();
+		}
+		EncryptCipherKeyDetailsRef cipherDetails = rep.cipherKeyDetails[0];
+		if (cipherDetails.encryptDomainId != SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID) {
+			TraceEvent(SevWarn, "KMSHealthCheckNoSystemKeyFound");
+			throw key_not_found();
+		}
+		CipherKeyValidityTS validityTS =
+		    getCipherKeyValidityTS(cipherDetails.refreshAfterSec, cipherDetails.expireAfterSec);
+		ekpProxyData->insertIntoBaseDomainIdCache(cipherDetails.encryptDomainId,
+		                                          cipherDetails.encryptKeyId,
+		                                          cipherDetails.encryptKey,
+		                                          cipherDetails.encryptKCV,
+		                                          validityTS.refreshAtTS,
+		                                          validityTS.expAtTS);
+		return KMSHealthStatus{ true, true, now() };
+	} catch (Error& e) {
+		TraceEvent(SevWarn, "KMSHealthCheckError").error(e);
+		if (!canReplyWith(e)) {
+			throw;
```

Comment:
do we want to report error in all cases? Also, what does it means by bubbling error up?

Reply:
Not in all cases, this check makes sure to not throw errors for everything in `canReplyWith`. For any other errors (i.e `actor_cancelled`) we will close the health check.

Reply:
This is similar logic for how our refresh keys actors work